### ISSUE TITLE
Make it possible to declare exclusions in profiles

### ISSUE THREAD

**Issue by kdsx** [conan-io/conan#4148]
Actually this is why I found #4135... (explained in the end)

Currently conan profiles allow glob expressions when specifying environment variables, build requirements, etc. However I haven't found any way to inverse the effect of a glob: apply to everything but the specified package.

Why I need this:
I have to build a project for a bunch of platforms including some devices like Raspberry Pi. Thus I need a cross-compiler. I build it using CrosstoolNg for a recipe was created. This recipe is a build dependency of another recipe for GCC toolchain. The resulting toolchain should be used to build almost everything else (including packages from the official conan repository, e.g. OpenSSL). Different devices require different options of GCC package (e.g. target libc version). To handle all this I created profiles for all supported platforms (e.g. linux-rpi.profile, linux-x64.profile, ...). All of them include some common profile (linux-common.profile) where I'd like to just write:
```
[build_requires]
gcc-toolchain/8.1@kds/test
```
Thus if I want to make a clean build of my project (including all dependencies) for RPI, I would just write:
```
$ conan install . -pr linux-rpi.profile --build=all
```
However this approach makes gcc-toolchain a dependency of itself, as well as a dependency of CrosstoolNg. Somehow CrosstoolNg survives this with no issue, but GCC is built two times. First time it is build with default options (from conanfile.py) and the result becomes the build requirement for the second GCC build which has correct options (from the profile). Then all other packages are build using the correct compiler. So the issue is strange double-build of GCC.

Basically the issue is not critical because I actually don't use `--build=all` (I wrote it for example), and the most of time GCC is pre-build and lays in the conan repository. However these self-dependencies look weird.

To avoid such effect I can enumerate all packages requiring cross-compilation in my common profile:
```
[build_requires]
ffmpeg*:gcc-toolchain/8.1@kds/test
libvpx*:gcc-toolchain/8.1@kds/test
...
```
But this is not a perfect solution because manual package enumeration is also weird, and also it currently suffers from the issue mentioned in the beginning (#4135). It's impossible to enter ffmpeg recipe directory and say `conan create . -pr ../profiles/linux-rpi.profile`. The dependency will not be used.

I'm not sure that the solution mentioned in the issue title is good because it will mess the whole profile with such exclusions just because of couple of packages. Probably it's better to make it possible to "redirect" a package to another profile.
E.g. something like this could say conan use the specified profile for GCC package and ignore everything declared in the current profile.
```
[override_profile]
gcc-toolchain*: linux-rpi-gcc.profile
```

**Comment by memsharded** [comment]
Hi @kdsx 

I am having a look at this issue. It would be great to know what dependency graph is conan evaluating for your use case. Maybe trying, in place of ``$ conan install . -pr linux-rpi.profile --build=all`` something like:

```bash
$ conan info . --pr=linux-rpi.profile --graph=file.html --dry-build
```
I think we should be able to see there that duplicated dependency.

It is true that the best approach right now is to build ``gcc-toolchain`` first, in a separate process. Not only because of this, but because the settings (the profile) you might be using to build the toolchain can be very different to the profile you are using to actually cross-build to the RPI. Lets say, you can build the toolchain itself with Visual Studio in Windows, but then it is going to be applied for gcc in Windows. To be able to model this complexity, Conan will be developing a new cross-build settings model, but that will take time. So pre-building ``gcc-toolchain`` is totally what makes sense here.

**Comment by kdsx** [comment]
Hi, @memsharded 

I've checked this again and got some details. Actually there's no issue when I build the whole project or something that depends on the toolchain. Sorry for wrong information. The issue appears when I build GCC itself using the profile.

When I enter gcc-toolchain recipe directory and run `$ conan create . kds/test -pr ../profiles/linux-rpi.profile`, I get the described self-dependency. The graph is attached.
![image](https://user-images.githubusercontent.com/10761841/50214944-aaf80400-0392-11e9-85f5-598874373949.png)

Thus I can make a clean build of the whole project. To build GCC separately, I can use a dedicated profile which will contain only basic settings and libc version option. So the issue exists, but is not critical.

**Comment by kdsx** [comment]
Also note the issue does not relate to a local recipe. The following commands also produce double-build:
 ```
$ cd gcc-toolchain
$ conan export . kds/test
$ conan install gcc-toolchain/8.1@kds/test -pr ../profiles/linux-rpi.profile
```

**Comment by memsharded** [comment]
Uhmmm, but are you building ``gcc-toolchain`` with a profile that says:

```
[build_requires]
gcc-toolchain/8.1@kds/test
```
?

Then it seems that it would be a wrong input, wouldn't it? You need to build ``gcc-toolchain`` with a profile without that build requires. Maybe the profiles composition with ``include(other_profile)`` could help you? Also, I am not fully sure about the other settings. For building ``gcc-toolchain`` you might actually use another compiler, with another compiler, another version and even another ``libcxx``. The profile for building ``gcc-toolchain`` to cross-build should contain the settings of the build machine, not the host (RPI) one.

**Comment by kdsx** [comment]
Yes, I build with such profile. I tried to explain why and why I wanted some kind of exclusion.

However, if there're plans to improve cross-compliling cases in conan, I do not insist on implementation of anythig I suggested. I'd like to see how conan will handle this first.

So far it's possible to live with it somehow :)

The only thing: maybe conan should complain in this situation instead of silently building something strange? Also strange that dependencies graph look depends on the root point. If I start with e.g. ffmpeg, I don't have wrong gcc-toolchain in the graph. If I start with gcc, it's there.

**Comment by memsharded** [comment]
Yes, I think that it totally makes sense to wait until the new cross-building model, because this is partially a symptom of that. I have assigned this issue to the project that manages this new model, and we will revisit when we start with it. Thanks!

**Comment by kdsx** [comment]
Thanks. Looking forward to it.

**Comment by kdsx** [comment]
I'd like to note that while this feature may be useful in certain scenarios, my original issue was resolved in 1.24 which introduced different profiles for host and build contexts (#5594). Thanks!

**Comment by memsharded** [comment]
> I'd like to note that while this feature may be useful in certain scenarios, my original issue was resolved in 1.24 which introduced different profiles for host and build contexts (#5594). Thanks!

Cool! The new model still lacks a couple of things, like making the build-helpers using them properly, and later migrating the ConanCenter packages to use it too. Please stay tuned for new releases and report possible issues. Thanks!

**Comment by memsharded** [comment]
Closing this as solved by the 2 profiles (build/host) cross-build model, released years ago in 1.X
